index.js

```diff
@@ -0,0 +1,414 @@
+ *   expected amount.
+ *
+ * @returns {boolean} True is the amounts equal.
+ */
+const compareAmount = ({order, expectedAmount}) => {
+  // get the total for the purchase_units.
+  const total = getTotalCost({order});
+  const sameAmount = total.isEqualTo(expectedAmount);
+  if(!sameAmount) {
+    throw new BedrockError(
+      `Expected ${expectedAmount} amount got ${total.toString()}`, Errors.Data);
+  }
+  return true;
+};
+
+/**
+ * Checks that the existing PayPal order is there
+ * and match the expectedAmount.
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.payment - A Bedrock payment.
+ * @param {object} options.amount - A Paypal amount object.
+ * @param {BigNumber|number|string} options.expectedAmount - How is expected.
+ *
+ * @returns {object} The updated paypal order.
+ */
+const updateGatewayPaymentAmount = async (
+  {payment, amount, expectedAmount}) => {
+  // ensure the order is still there.
+  const order = await getOrderFromPayment({payment});
+  const value = formatAmount({amount});
+  compareAmount({order, expectedAmount});
+  const patch = [{
+    op: 'replace',
+    path: `/purchase_units/@reference_id=='${payment.id}'/amount`,
+    value
+  }];
+  const updatedOrder = await updateOrder({order, patch});
+  // make sure the swap occurred.
+  compareAmount({order: updatedOrder, expectedAmount: amount.value});
+  return updatedOrder;
+};
+
+/**
+ * This first checks to ensure the client's order matches
+ * a valid PayPal Order.
+ *
+ * @param {object} options - Options to use.
+ * @param {Array<object>} options.clientPurchases - Client purchases.
+ * @param {Array<object>} options.paypalPurchases - PayPal purchases.
+ *
+ * @returns {object} A verified purchase.
+ */
+const verifyPurchase = ({paypalPurchases}) => {
+  // PayPal currently only supports one purchase_item at a time.
+  // Hence we use the first purchase_item.
+  // This will need to be refactored if we allow a user
+  // to purchase more than one product in an order or PayPal allows
+  // more than one purchase_unit per order.
+  //
+  if(!paypalPurchases) {
+    throw new BedrockError('Missing PayPal purchases', Errors.Data);
+  }
+  const [paypalPurchase] = paypalPurchases;
+  if(!paypalPurchase) {
+    throw new BedrockError('Missing PayPal purchase', Errors.Data);
+  }
+  // throw an error if the client altered their purchase amount
+  // after the transaction succeeded.
+  return paypalPurchase;
+};
+
+/**
+ * Gets the order from PayPal then ensures the client's order
+ * matches the order from PayPal.
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.order - Client's PayPal order.
+ *
+ * @returns {Promise<object>} Returns the verified PayPal order and purchase.
+ */
+const verifyOrder = async ({order}) => {
+  // An order should never be CREATED
+  // at this stage so let's just cancel it.
+  if(order.status === 'CREATED') {
+    await deleteOrder({order});
+    return {};
```

Comment:
Seems like this is some unexpected state and an error should be thrown after doing the cleanup.

Reply:
this propagates up to the HTTP Handler which then throws there. I'm not sure what I think about this design either.

Reply:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/afa1265a02cffe57b8c23a27a2045e62c4b1f466

I add `void: true` to details and then in bedrock-payment-http if an error has `void:true` in details we void the payment there.

Reply:
Just so you know this was further revised. I pass payment into verifyOrder now so I can set the payment.status inside of the plugin and not in bedrock-payment-http.